File: src/main/java/main/json/Json.java
package main.json;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 *
 * @author Bunnyspa
 */
public interface Json {

    int NULL = 0;
    int BOOLEAN = 1;
    int NUMBER = 2;
    int TEXT = 3;
    int ARRAY = 4;
    int OBJECT = 5;

    int getType();

    @NotNull
    static Json parse(@NotNull String data) {
        String dataTrim = data.trim();
        switch (dataTrim.charAt(0)) {
            // Object
            case '{':
                return new ObjectJson(data);
            // Array
            case '[':
                return new ArrayJson(data);
            // Text
            case '"':
                return new TextJson(data);
            // Boolean
            case 't':
            case 'f':
                return new BooleanJson(data);
            // Null
            case 'n':
                return new NullJson();
            // Number
            default:
                return new NumberJson(data);
        }
    }

    static int getEndIndex(@NotNull String data, int i) {
        while (i < data.length()) {
            char next = data.charAt(i);
            if (!Character.isWhitespace(next)) {
                switch (next) {
                    // Object
                    case '{':
                        return getBracketEndIndex(data, i, OBJECT);
                    // Array
                    case '[':
                        return getBracketEndIndex(data, i, ARRAY);
                    // Text
                    case '"':
                        i++;
                        while (i < data.length()) {
                            if (data.charAt(i) == '"' && (i == 0 || data.charAt(i - 1) != '\\')) {
                                return i;
                            }
                            i++;
                        }
                        return -1;
                    // Boolean - True
                    case 't':
                        if (data.startsWith("true", i)) {
                            return i + 4;
                        }
                        return -1;
                    // Boolean - False
                    case 'f':
                        if (data.startsWith("false", i)) {
                            return i + 5;
                        }
                        return -1;
                    // Null
                    case 'n':
                        if (data.startsWith("null", i)) {
                            return i + 4;
                        }
                        return -1;
                    // Number
                    default:
                        i++;
                        while (i < data.length()) {
                            char c = data.charAt(i);
                            if (c != '.' && (c < '0' || '9' < c)) {
                                return i - 1;
                            }
                            i++;
                        }
                        return i - 1;
                }
            }
            i++;
        }
        return -1;
    }

    static int getBracketEndIndex(@NotNull String data, int i, int type) {
        try {
            int bracketLevel = 0;
            boolean quoting = false;

            char openBracket = type == OBJECT ? '{' : '[';
            char closeBracket = type == OBJECT ? '}' : ']';

            do {
                char c = data.charAt(i);
                if (c == '"' && (i == 0 || data.charAt(i - 1) != '\\')) {
                    quoting = !quoting;
                }
                if (!quoting && c == openBracket) {
                    bracketLevel++;
                }
                if (!quoting && c == closeBracket) {
                    bracketLevel--;
                }
                i++;
            } while (bracketLevel > 0 && i < data.length());
            return i - 1;
        } catch (Exception ex) {
            return -1;
        }
    }

    static boolean getBoolean(@NotNull Json data) throws ClassCastException {
        if (data.getType() != BOOLEAN) {
            throw new ClassCastException(getClassCastExceptionMessage(data, BooleanJson.class.getName()));
        }
        return ((BooleanJson) data).getBoolean();
    }

    static int getInteger(@NotNull Json data) throws ClassCastException {
        if (data.getType() != NUMBER) {
            throw new ClassCastException(getClassCastExceptionMessage(data, NumberJson.class.getName()));
        }
        return ((NumberJson) data).getInteger();
    }

    static double getDouble(@NotNull Json data) throws ClassCastException {
        if (data.getType() != NUMBER) {
            throw new ClassCastException(getClassCastExceptionMessage(data, NumberJson.class.getName()));
        }
        return ((NumberJson) data).getDouble();
    }

    @NotNull
    static String getText(@NotNull Json data) throws ClassCastException {
        if (data.getType() != TEXT) {
            throw new ClassCastException(getClassCastExceptionMessage(data, TextJson.class.getName()));
        }
        return ((TextJson) data).getText();
    }

    @NotNull
    static List<Json> getList(@NotNull Json data) throws ClassCastException {
        if (data.getType() != ARRAY) {
            throw new ClassCastException(getClassCastExceptionMessage(data, ArrayJson.class.getName()));
        }
        return ((ArrayJson) data).getList();
    }

    @NotNull
    static ObjectJson getObjectJson(@NotNull Json data) throws ClassCastException {
        if (data.getType() != OBJECT) {
            throw new ClassCastException(getClassCastExceptionMessage(data, ObjectJson.class.getName()));
        }
        return (ObjectJson) data;
    }

    @NotNull
    static List<String> getObjectKeys(@NotNull Json data) throws ClassCastException {
        if (data.getType() != OBJECT) {
            throw new ClassCastException(getClassCastExceptionMessage(data, ObjectJson.class.getName()));
        }
        return ((ObjectJson) data).getKeys();
    }

    @Nullable
    static Json getObjectValue(@NotNull Json data, String key) throws ClassCastException {
        if (data.getType() != OBJECT) {
            throw new ClassCastException(getClassCastExceptionMessage(data, ObjectJson.class.getName()));
        }
        return ((ObjectJson) data).getValue(key);
    }

    @NotNull
    static String getClassCastExceptionMessage(@NotNull Json data, String cast) {
        StringBuilder sb = new StringBuilder();
        sb.append(Json.class.getName()).append(" cannot be cast to ").append(cast).append("- It should be cast to ");
        switch (data.getType()) {
            case NULL:
                sb.append(NullJson.class.getName());
                break;
            case BOOLEAN:
                sb.append(BooleanJson.class.getName());
                break;
            case NUMBER:
                sb.append(NumberJson.class.getName());
                break;
            case TEXT:
                sb.append(TextJson.class.getName());
                break;
            case ARRAY:
                sb.append(ArrayJson.class.getName());
                break;
            case OBJECT:
                sb.append(ObjectJson.class.getName());
                break;
            default:
                throw new AssertionError();
        }
        return sb.toString();
    }
}

File: src/main/java/main/json/Json.kt
package main.json

/**
 *
 * @author Bunnyspa
 */
interface Json {
    val type: Int

    companion object {
        fun parse(data: String): Json {
            val dataTrim = data.trim { it <= ' ' }
            return when (dataTrim[0]) {
                '{' -> ObjectJson(data)
                '[' -> ArrayJson(data)
                '"' -> TextJson(data)
                't', 'f' -> BooleanJson(data)
                'n' -> NullJson()
                else -> NumberJson(data)
            }
        }

        fun getEndIndex(data: String, i: Int): Int {
            var i = i
            while (i < data.length) {
                val next = data[i]
                if (!Character.isWhitespace(next)) {
                    return when (next) {
                        '{' -> getBracketEndIndex(data, i, OBJECT)
                        '[' -> getBracketEndIndex(data, i, ARRAY)
                        '"' -> {
                            i++
                            while (i < data.length) {
                                if (data[i] == '"' && (i == 0 || data[i - 1] != '\\')) {
                                    return i
                                }
                                i++
                            }
                            -1
                        }
                        't' -> {
                            if (data.startsWith("true", i)) {
                                i + 4
                            } else -1
                        }
                        'f' -> {
                            if (data.startsWith("false", i)) {
                                i + 5
                            } else -1
                        }
                        'n' -> {
                            if (data.startsWith("null", i)) {
                                i + 4
                            } else -1
                        }
                        else -> {
                            i++
                            while (i < data.length) {
                                val c = data[i]
                                if (c != '.' && (c < '0' || '9' < c)) {
                                    return i - 1
                                }
                                i++
                            }
                            i - 1
                        }
                    }
                }
                i++
            }
            return -1
        }

        private fun getBracketEndIndex(data: String, i: Int, type: Int): Int {
            var i = i
            return try {
                var bracketLevel = 0
                var quoting = false
                val openBracket = if (type == OBJECT) '{' else '['
                val closeBracket = if (type == OBJECT) '}' else ']'
                do {
                    val c = data[i]
                    if (c == '"' && (i == 0 || data[i - 1] != '\\')) {
                        quoting = !quoting
                    }
                    if (!quoting && c == openBracket) {
                        bracketLevel++
                    }
                    if (!quoting && c == closeBracket) {
                        bracketLevel--
                    }
                    i++
                } while (bracketLevel > 0 && i < data.length)
                i - 1
            } catch (ex: Exception) {
                -1
            }
        }

        @Throws(ClassCastException::class)
        fun getText(data: Json): String {
            if (data.type != TEXT) {
                throw ClassCastException(getClassCastExceptionMessage(data, TextJson::class.java.name))
            }
            return (data as TextJson).text
        }

        @Throws(ClassCastException::class)
        fun getList(data: Json): List<Json?> {
            if (data.type != ARRAY) {
                throw ClassCastException(getClassCastExceptionMessage(data, ArrayJson::class.java.name))
            }
            return (data as ArrayJson).list
        }

        @Throws(ClassCastException::class)
        fun getObjectJson(data: Json?): ObjectJson {
            if (data?.type != OBJECT) {
                throw ClassCastException(getClassCastExceptionMessage(data, ObjectJson::class.java.name))
            }
            return data as ObjectJson
        }

        @Throws(ClassCastException::class)
        fun getObjectKeys(data: Json): List<String?> {
            if (data.type != OBJECT) {
                throw ClassCastException(getClassCastExceptionMessage(data, ObjectJson::class.java.name))
            }
            return (data as ObjectJson).keys
        }

        @Throws(ClassCastException::class)
        fun getObjectValue(data: Json, key: String?): Json? {
            if (data.type != OBJECT) {
                throw ClassCastException(getClassCastExceptionMessage(data, ObjectJson::class.java.name))
            }
            return (data as ObjectJson).getValue(key)
        }

        private fun getClassCastExceptionMessage(data: Json?, cast: String?): String {
            val sb = StringBuilder()
            sb.append(Json::class.java.name).append(" cannot be cast to ").append(cast)
                .append("- It should be cast to ")
            when (data?.type) {
                NULL -> sb.append(NullJson::class.java.name)
                BOOLEAN -> sb.append(BooleanJson::class.java.name)
                NUMBER -> sb.append(NumberJson::class.java.name)
                TEXT -> sb.append(TextJson::class.java.name)
                ARRAY -> sb.append(ArrayJson::class.java.name)
                OBJECT -> sb.append(ObjectJson::class.java.name)
                else -> throw AssertionError()
            }
            return sb.toString()
        }

        const val NULL = 0
        const val BOOLEAN = 1
        const val NUMBER = 2
        const val TEXT = 3
        const val ARRAY = 4
        const val OBJECT = 5
    }
}